scene/resources/particles_material.cpp

```diff
@@ -282,6 +291,73 @@ void ParticlesMaterial::_update_shader() {
 		code += "uniform float collision_bounce;\n";
 	}
 
+	if (turbulence_mode != 0) {
+		code += "uniform float turbulence_noise_contrast;\n";
+		code += "uniform float turbulence_noise_scale;\n";
+		code += "uniform float turbulence_position_influence;\n";
+		code += "uniform float turbulence_velocity_influence;\n";
+		code += "uniform float turbulence_noise_speed_random;\n";
+		code += "uniform vec3 turbulence_noise_speed = vec3(1.0, 1.0, 1.0);\n";
+		if (tex_parameters[PARAM_TURB_INFLUENCE].is_valid()) {
+			code += "uniform sampler2D turbulence_influence_over_life;\n";
+		}
+		if (turbulence_color_ramp.is_valid()) {
+			code += "uniform sampler2D turbulence_color_ramp;\n";
+		}
+		code += "\n";
+
+		//functions for 3D noise / turbulence
+		code += "\n\n";
+		code += "// 3D Noise with friendly permission by Inigo Quilez\n";
+		code += "vec3 hash_noise( vec3 p ) {\n";
```

Comment:
Just a short comment for when you are back to working on this PR. We should use a cheaper hash function and one that will behave better over space. The current one (while slightly improved by removing the sin) breaks down pretty quickly over time.

I suggest "Hash without Sin" By Dave Hoskins on Shadertoy: https://www.shadertoy.com/view/4djSRW
```
vec3 hash33(vec3 p3) {
    p3 = fract(p3 * vec3(.1031, .1030, .0973));
    p3 += dot(p3, p3.yxz+33.33);
    return fract((p3.xxy + p3.yxx)*p3.zyx);
}
```
This is much cheaper and it behaves well over a very large spatial domain. In the linked shadertoy you can see a visual comparison between the current hash function and this one that I am proposing.